### ISSUE TITLE
[CUDA] illegal memory Write on _convert_indices_from_coo_to_csr

### ISSUE THREAD

**Issue by kokol16** [pytorch/pytorch#182064]
### 🐛 Describe the bug

```python
import torch
print(torch.__version__)
torch.cuda.init()

gpu_dev = torch.device('cuda')

tensor_0 = torch.full((1,), 2147480000, dtype=torch.int64, requires_grad=False, device=gpu_dev)
int_1 = -1
bool_2 = False
tensor_3 = torch.full((1,), 2147480000, dtype=torch.int64, requires_grad=False, device=gpu_dev)
torch.ops.aten._convert_indices_from_coo_to_csr(tensor_0, int_1, out_int32=bool_2)
torch.cuda.synchronize()
```
```
========= COMPUTE-SANITIZER
========= Invalid __global__ write of size 8 bytes
=========     at void at::native::<unnamed>::convert_indices_from_coo_to_csr_cuda_kernel<long, long>(T2 *, const T1 *, long, long)+0xa80
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x0 is out of bounds
=========     and is 140604711370752 bytes before the nearest allocation at 0x7fe115e00000 of size 2097152 bytes
=========     Saved host backtrace up to driver entry point at kernel launch time
```

### Versions

Collecting environment information...
PyTorch version: 2.11.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Mar  3 2026, 11:56:32) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-174-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA A30
GPU 1: NVIDIA A30

Nvidia driver version: 560.28.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.28.9
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.11.0+cu126
[pip3] torchaudio==2.11.0+cu126
[pip3] torchvision==0.26.0+cu126
[pip3] triton==3.6.0
[pip3] tritonclient==2.67.0
[conda] _anaconda_depends         2023.09             py311_mkl_1  
[conda] blas                      1.0                         mkl  
[conda] intel-openmp              2023.1.0         hdb19cb5_46306  
[conda] magma-cuda121             2.6.1                         1    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-include               2023.2.0            intel_49495    intel
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl-static                2023.2.0            intel_49495    intel
[conda] mkl_fft                   1.3.8           py311h5eee18b_0  
[conda] mkl_random                1.2.4           py311hdb19cb5_0  
[conda] numpy                     1.26.4          py311h08b1b3b_0  
[conda] numpy-base                1.26.4          py311hf175353_0  
[conda] numpydoc                  1.5.0           py311h06a4308_0  
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] pytorch-triton            3.0.0+989adb9a29          pypi_0    pypi
[conda] tbb                       2021.8.0             hdb19cb5_0  
[conda] tbb4py                    2021.8.0        py311hdb19cb5_0  
[conda] torch                     2.3.0a0+git443444d           dev_0    <develop>

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia

**Comment by pearu** [comment]
This result is not specific to CUDA as on CPU the example would crash as well.

Also, this issue is not actionable: if one uses invalid input (`size=-1` in this case) to an internal function such as `_convert_indices_from_coo_to_csr` that is optimized to not deal with checking invalid inputs, the crash is expected.

**Comment by Arindamsamanta2004** [comment]
Hi @pearu and @drisspg,

I've been investigating this issue and wanted to get your input before proceeding with any changes.

The current situation is the function `_convert_indices_from_coo_to_csr` is an internal function that currently,
- Validates input tensor dimensions (`TORCH_CHECK(self.dim() <= 1, ...)`)
- Does NOT validate the `size` parameter

When `size = -1`:
1. Output tensor is allocated with `size + 1 = 0` elements 
2. CUDA kernel tries to write to indices `[0...2147480000]` in this zero-sized buffer
3. Result is illegal memory write at address 0x0

@pearu mentioned this is expected behavior for an internal function optimized to skip validation. I'm trying to understand this perspective and I can see it's consistent with other `_unsafe` functions like `_sparse_csr_tensor_unsafe()`.

However, I also noticed that:

1. `_convert_indices_from_coo_to_csr` already validates one parameter (tensor dimensions)                                                                                                                                                             
2. All internal callers of this function pass valid values (e.g., `self.size(0)`)
3. Adding `TORCH_CHECK(size >= 0, ...)` is just a single integer comparison

Our options are to add runtime validation or a debug-only validation.

Since all internal callers already pass valid size values, this issue is fuzzer-only. I'm happy to send a PR with the option that aligns or we can close the issue if you prefer.

**Comment by pearu** [comment]
Thanks for the careful investigation, @Arindamsamanta2004 — your reasoning about the existing `dim` check and the negligible cost is sound, and I agree we should add the guard.

Let me state the general principle we apply to these internal conversion ops, since it clarifies exactly which checks belong here:

- **O(1) host-side checks are fine** (and welcome). A scalar sanity check like `size >= 0` runs once on the host before any kernel launch, costs nothing measurable, and converts UB into a clean error. The existing `self.dim() <= 1` check is exactly this kind, so adding `size >= 0` next to it is the consistent thing to do.
- **Checks that require reading the data are not.** Validating input *values* means a device reduction and a sync. For these ops we deliberately minimize kernel/sync calls and instead rely on the caller to provide correct data. So we do not add them.

Concretely: I'd take a runtime `TORCH_CHECK(size >= 0, ...)` (not debug-only) in `TORCH_META_FUNC(_convert_indices_from_coo_to_csr)`, right next to the `dim` check.

The one thing I want to set expectations on: `size >= 0` fixes the reported case but does **not** make the op memory-safe in general. The actual invariant the kernel relies on is `max(self) <= size`, which falls squarely in the second category above — it would need a sync, so we won't validate it. Example that survives the `size >= 0` guard:

```python
import torch
# crow buffer has size+1 = 2 slots, but the kernel writes data_out[0..5]
torch.ops.aten._convert_indices_from_coo_to_csr(torch.tensor([5]), 1)
```

The first loop in `convert_indices_from_coo_to_csr_cpu` writes `data_out[0 .. data_in[0]]`, and nothing ties `data_in` values to `size`. All internal callers pass `self.size(0)`, which satisfies `max(self) <= size` by construction.

So, to be precise about scope: the PR converts the `size < 0` case from UB (illegal CUDA write / CPU heap overflow) into a clean `RuntimeError`, while the op remains "trust the caller" for the index-vs-size relationship. A PR adding the check is welcome — please include a test that reproduces this issue (the `size < 0` case) and asserts the new error.

**Comment by Arindamsamanta2004** [comment]
@pearu Thanks for the clarification! The PR adds `TORCH_CHECK(size >= 0, ...)` in the meta function right next to the dim check, along with a test case for the negative size scenario. Do let me know if this looks good?
